### PR TITLE
Network2

### PR DESCRIPTION
**Pull request by brianolson**
### Description

Huge refactor of network code.

**Design Wins**
 * Inbound message used to hop across 4-5 queues end to end, now it’s one
 * Code used to send command messages to internal server processes, now it’s just library code that runs in app thread or peer thread
 * ( _to be continued..._ )

### CODE REVIEW

File: network/framework/src/connectivity_manager/mod.rs [hunk 1]
```diff
@@ -436,47 +482,43 @@ where
     /// reconfiguration. If we are currently connected to this validator, calling
     /// this function will close our connection to it.
     async fn close_stale_connections(&mut self) {
```
Comment (brianolson) [inline network/framework/src/connectivity_manager/mod.rs:484]: something is wrong with this implementation of 'stale'. I think that 'stale' is supposed to be "used to be in the Validator set", so "if (this is the Validator network) AND (a node is not a validator) { close connection to it }" but actually this logic seems to close most of the connections most of the time.

File: aptos-node/src/lib.rs [hunk 2]
```diff
@@ -599,22 +604,50 @@ pub fn setup_environment_and_start_node(
...
+    let dkg_network_interfaces = network2::dkg_network_connections(&mut apps, &app_setup);
+    let jwk_consensus_network_interfaces = network2::jwk_consensus_network_connections(&mut apps, &app_setup);
+    let netbench_network_interfaces = network2::netbench_network_connections(&mut apps, &app_setup);
+
+    for (protocol_id, ac) in apps.iter() {
+        info!("app_int setup {} -> {} {:?}", protocol_id.as_str(), ac.label, &ac.sender);
+    }
+    let apps = Arc::new(apps);
+    for network in networks.iter_mut() {
+        network.set_apps(apps.clone());
```
Comment (JoshLind) [inline aptos-node/src/lib.rs:648]: We set all apps for all network types? Traditionally, apps are specific to network type, no? e.g., consensus only operates in the validator network, etc.

File: network/benchmark/src/lib.rs [hunk 3]
```diff
@@ -273,57 +253,175 @@ pub async fn run_netbench_service(
...
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                ));
+            }
+        }
+    }
+    let mut connection_notifications = network_client.peers_and_metadata.subscribe();
```
Comment (JoshLind) [inline network/benchmark/src/lib.rs:320]: Interesting... I guess a subscription model means that we don't unnecessarily need to keep looping over the existing set (to identify new and lost peers)?

File: network/builder/src/lib.rs [hunk 4]
```diff
@@ -1,14 +1,299 @@
...
+        nb.setup_discovery(reconfig_subscription_service, connectivity_req_sender);
+        nb
+    }
+
+    pub fn set_apps(&mut self, apps: Arc<ApplicationCollector>) {
+        self.apps = apps;
+    }
+
+    pub fn active_protocol_ids(&self) -> ProtocolIdSet {
+        let mut out = ProtocolIdSet::empty();
```
Comment (JoshLind) [inline network/builder/src/lib.rs:109]: nit: `out` -> `protocol_ids`? 😄  (I know, it's a style thing...) 😆

File: network/builder/src/lib.rs [hunk 5]
```diff
@@ -1,14 +1,299 @@
...
+    Mutual(x25519::PrivateKey),
+}
+
+/// Build Network module with custom configuration values.
+/// Methods can be chained in order to set the configuration values.
+/// MempoolNetworkHandler and ConsensusNetworkHandler are constructed by calling
+/// [`NetworkBuilder::build`].  New instances of `NetworkBuilder` are obtained
+/// via [`NetworkBuilder::create`].
+pub struct NetworkBuilder {
+    state: State,
```
Comment (JoshLind) [inline network/builder/src/lib.rs:59]: Why exactly do we need this?

File: state-sync/storage-service/server/src/handler.rs [hunk 6]
```diff
@@ -255,14 +255,16 @@ impl<T: StorageReaderInterface> Handler<T> {
                     .request(&request)
                 );
             );
-        }
+            metrics::OPTIMISTIC_FETCH_EVENTS.with_label_values(&[peer_network_id.network_id().as_str(), "collision"]).inc();
```
Comment (JoshLind) [inline state-sync/storage-service/server/src/handler.rs:258]: These could be in a separate PR 😄 Also, not entirely sure about the value of the metric, e.g., we expect this to happen, but I'm not sure if the frequency means much.

File: network/framework/src/setup.rs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+use tokio::runtime::Runtime;
+use aptos_config::config::{NetworkConfig, NodeConfig};
+use aptos_event_notifications::EventSubscriptionService;
+use aptos_logger::debug;
+use aptos_time_service::TimeService;
+use aptos_types::chain_id::ChainId;
+use crate::application::storage::PeersAndMetadata;
+// use crate::builder::NetworkBuilder;
+
+pub fn setup_applications() {
```
Comment (JoshLind) [inline network/framework/src/setup.rs:14]: ? 😄

File: network/framework/src/protocols/rpc_error.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_types::PeerId;
+use futures::channel::{mpsc, oneshot};
+use std::io;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum RpcError {
```
Comment (JoshLind) [inline network/framework/src/protocols/rpc_error.rs:11]: Nice. To make this easier to land, you could add these types of refactors to a separate PR 😄 Will be easier to review.

File: network/framework/src/protocols/rpc/error.rs [hunk 9]
```diff
@@ -37,22 +35,16 @@ pub enum RpcError {
     #[error("Error sending on mpsc channel, connection likely shutting down: {0:?}")]
     MpscSendError(#[from] mpsc::SendError),
 
+    #[error("Error sending on mpsc channel, connection likely shutting down")]
+    TokioMpscSendError,
```
Comment (JoshLind) [inline network/framework/src/protocols/rpc/error.rs:39]: Do we need a separate error for this? Why is the above insufficient?

File: network/framework/src/protocols/health_checker/test.rs [hunk 10]
```diff
@@ -215,11 +215,11 @@ async fn outbound() {
         // Health Checker should attempt to ping the new peer.
         harness.expect_ping_send_ok().await;
     };
-    future::join(health_checker.start(), test).await;
+    block_on(future::join(health_checker.start(), test));
```
Comment (JoshLind) [inline network/framework/src/protocols/health_checker/test.rs:218]: Curious why you changed this? 😄

**Review by JoshLind — COMMENTED** [review]
Some random comments. Got about half way through 😄 Will return to the meat. But, I've added some comments.

File: network/framework/src/application/interface.rs [hunk 11]
```diff
@@ -221,24 +245,141 @@ impl<Message: NetworkMessageTrait> NetworkClientInterface<Message> for NetworkCl
...
+
+impl Default for OutboundRpcMatcher {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub fn protocol_is_high_priority(protocol_id: ProtocolId) -> bool {
+    match protocol_id {
+        ProtocolId::ConsensusRpcBcs => {true}
```
Comment (JoshLind) [inline network/framework/src/application/interface.rs:368]: nit: you could match multiple items at a time, e.g.,:
```
match protocol_id {
    ProtocolId::ConsensusRpcBcs | ProtocolId::ConsensusDirectSendBcs | .... => true,
    _ => false,
}
```

**Review by JoshLind — COMMENTED** [review]
One big question on my mind is how we plan to land this once we've tested it. It's fine as-is for `previewnet`, but it's much too large a change for me to reason about at once.

Looking at the PR, it seems like there's a number of changes that I'm hoping can be submitted independently. But, want to see how you are thinking 😄

File: network/framework/src/noise/handshake.rs [hunk 12]
```diff
@@ -74,64 +74,6 @@ impl AntiReplayTimestamps {
     }
 }
 
-/// Noise handshake authentication mode.
-pub enum HandshakeAuthMode {
```
Comment (JoshLind) [inline network/framework/src/noise/handshake.rs:78]: If I understand this change correctly, you could probably separate this into a different commit?

File: network/framework/src/noise/handshake.rs [hunk 13]
```diff
@@ -165,13 +108,38 @@ impl NoiseUpgrader {
...
+            peers_and_metadata,
+            anti_replay_timestamps,
+        }
+    }
+
+    pub fn is_mutual_mode(&self) -> bool {
+        self.anti_replay_timestamps.is_some()
+    }
+
+    pub fn is_trusted_peer(&self, network_id: &NetworkId, peer_id: &PeerId) -> Option<Peer> {
```
Comment (JoshLind) [inline network/framework/src/noise/handshake.rs:131]: nit: `is_trusted_peer` makes me want to expect a `bool` 😄

File: network/framework/src/application/storage.rs [hunk 14]
```diff
@@ -362,34 +459,82 @@ impl PeersAndMetadata {
...
-fn missing_peer_metadata_error(peer_network_id: &PeerNetworkId) -> Error {
+fn missing_metadata_error(peer_network_id: &PeerNetworkId) -> Error {
     Error::UnexpectedError(format!(
         "No metadata was found for the given peer: {:?}",
         peer_network_id
     ))
 }
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
+pub enum DisconnectReason {
```
Comment (JoshLind) [inline network/framework/src/application/storage.rs:488]: I'm still not entirely sure we need to keep this around. If we're adding generational support to the peers and metadata, surely it's cheap enough for applications to just keep polling the active peer set to be notified of changes?

File: network/framework/src/connectivity_manager/mod.rs [hunk 15]
```diff
@@ -482,47 +525,47 @@ where
...
-                        disconnect_error
-                    );
+                match self.peer_senders.get_generational(self.peer_senders_generation) {
+                    None => {}
+                    Some((new_peer_senders, new_generation)) => {
+                        self.peer_senders_cache = new_peer_senders;
+                        self.peer_senders_generation = new_generation;
+                    }
+                }
+                #[cfg(disabled)] // TODO: actually closing 'stale' is disabled until fixed
```
Comment (JoshLind) [inline network/framework/src/connectivity_manager/mod.rs:552]: Were there only specific types of connections being closed? Or all of them? My first guess is that somewhere the roles might have been switched up 🤔

File: network/framework/src/connectivity_manager/mod.rs [hunk 16]
```diff
@@ -985,75 +1122,57 @@ where
...
-                .set_trusted_peers(&self.network_context.network_id(), new_eligible)
-            {
-                error!(
-                    NetworkSchema::new(&self.network_context),
-                    error = %error,
-                    "Failed to update trusted peers set"
-                );
+            // Swap in the new eligible peers set. Drop the old set after releasing
+            // the write lock.
+            if let Some(trusted_peers) = self.get_trusted_peers() {
```
Comment (JoshLind) [inline network/framework/src/connectivity_manager/mod.rs:1127]: nit: why this change?

File: network/framework/src/peer.rs [hunk 17]
```diff
@@ -0,0 +1,861 @@
...
+use aptos_metrics_core::{IntCounter, IntCounterVec, register_int_counter_vec};
+use crate::application::ApplicationCollector;
+use crate::application::interface::{OpenRpcRequestState, OutboundRpcMatcher};
+use crate::application::storage::PeersAndMetadata;
+use crate::ProtocolId;
+use crate::protocols::network::{Closer, OutboundPeerConnections, PeerStub, ReceivedMessage};
+use crate::protocols::stream::{StreamFragment, StreamHeader, StreamMessage};
+use crate::transport::ConnectionMetadata;
+use once_cell::sync::Lazy;
+#[cfg(test)]
```
Comment (JoshLind) [inline network/framework/src/peer.rs:25]: nit: if you define a test module below, it should make all of these imports cleaner.

**Review by JoshLind — COMMENTED** [review]
Some additional comments 😄 Still not done but it's becoming apparent that there's a lot baked in here. We'll need to find a way to break things up when we come to land things in `main` 😃

File: network/framework/src/application/storage.rs (hunk 14 above)
Comment (brianolson) [reply network/framework/src/application/storage.rs:488]: things _want_ events. but maybe I should look again because maybe _tests_ want events, and they can go fish if that's not what's right for the application code. But if things wind up polling and diffing to create events, we should just provide events. I think the event model is at least cleaner now that it's out of band with content messages and there's a separate Subscribe flow for connect/disconnect events.

File: network/framework/src/protocols/stream/mod.rs [hunk 18]
```diff
@@ -18,6 +15,22 @@ pub enum StreamMessage {
...
+impl StreamMessage {
+    pub fn data_len(&self) -> usize {
+        match self {
+            StreamMessage::Header(head) => {head.message.data_len()}
+            StreamMessage::Fragment(frag) => {frag.raw_data.len()}
+        }
+    }
+    pub fn header_len(&self) -> usize {
+        match self {
+            StreamMessage::Header(head) => {head.message.header_len() + 5}
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/protocols/stream/mod.rs:27]: magic knowledge of the BCS stuff that's happening in the NetworkMessage wrapping the StreamMessage

File: network/framework/src/noise/handshake.rs (hunk 12 above)
Comment (brianolson) [reply network/framework/src/noise/handshake.rs:78]: Yeah, that's one of the earlier things I did and possibly separable

File: network/framework/src/noise/handshake.rs (hunk 13 above)
Comment (brianolson) [reply network/framework/src/noise/handshake.rs:131]: renaming to `peer_config_if_trusted()`

File: aptos-node/src/lib.rs [hunk 19]
```diff
@@ -599,22 +604,51 @@ pub fn setup_environment_and_start_node(
...
+    let contexts = Arc::new(contexts);
+
+    let app_setup = AppSetupContext{
+        node_config: node_config.clone(),
+        peers_and_metadata: peers_and_metadata.clone(),
+        peer_senders,
+        contexts,
+    };
+
+    let health_checker_network_interfaces = network2::health_checker_network_connections(&mut apps, &app_setup);
```
Comment (JoshLind) [inline aptos-node/src/lib.rs:635]: I imagine you could roll up a lot of this new connectivity functionality into a separate PR? Even if we don't move to the "new interfaces" entirely, we could adopt some of the patterns in the existing code?

File: config/src/config/network_config.rs [hunk 20]
```diff
@@ -343,6 +344,46 @@ impl NetworkConfig {
         }
         Ok(())
     }
+
+    /// Retrieve and merge seeds so that they have all keys associated
```
Comment (JoshLind) [inline config/src/config/network_config.rs:348]: Seems like a simple file move/refactor, right? This could be a separate PR 😄

File: consensus/src/block_storage/tracing.rs [hunk 21]
```diff
@@ -20,7 +20,7 @@ impl BlockStage {
     pub const QC_AGGREGATED: &'static str = "qc_aggregated";
     pub const RAND_ADD_DECISION: &'static str = "rand_add_decision";
     pub const RAND_ADD_SHARE: &'static str = "rand_add_share";
-    pub const RAND_AGG_DECISION: &'static str = "rand_agg_decision";
+    // pub const RAND_AGG_DECISION: &'static str = "rand_agg_decision";
```
Comment (JoshLind) [inline consensus/src/block_storage/tracing.rs:23]: Why did we remove this?

File: network/framework/src/application/storage.rs [hunk 22]
```diff
@@ -16,99 +16,145 @@ use aptos_config::{
...
-        Ok(all_peers)
+        all_peers
     }
 
-    /// Returns metadata for all peers currently connected to the node
-    pub fn get_connected_peers_and_metadata(
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_generational(
```
Comment (JoshLind) [inline network/framework/src/application/storage.rs:101]: I imagine the generational changes could be separated out?

File: network/framework/src/application/storage.rs [hunk 23]
```diff
@@ -117,55 +163,135 @@ impl PeersAndMetadata {
...
+        for evict in to_del.into_iter() {
+            let llast = listeners.len() - 1;
+            if evict != llast {
+                listeners.swap(evict, llast);
+            }
+            listeners.pop();
+        }
+    }
+
+    pub fn subscribe(&self) -> tokio::sync::mpsc::Receiver<ConnectionNotification> {
```
Comment (JoshLind) [inline network/framework/src/application/storage.rs:283]: Likewise, the new subscription service could be separated out?

File: network/framework/src/application/tests.rs [hunk 24]
```diff
@@ -583,247 +518,247 @@ async fn test_network_client_missing_network_sender() {
...
-        &mut vfn_network_events,
-        true,
-        None,
-        Some(ProtocolId::ConsensusRpcJson),
-        dummy_message,
-    )
-    .await;
-}
+// #[tokio::test]
+// async fn test_network_client_senders_no_matching_protocols() { // TODO network2 recreate something like this
```
Comment (JoshLind) [inline network/framework/src/application/tests.rs:522]: Yeah, we'll probably need to recreate these for the new interfaces 😄

File: network/framework/src/counters.rs [hunk 25]
```diff
@@ -39,22 +41,22 @@ const PRE_DIAL_LABEL: &str = "pre_dial";
 pub const SERIALIZATION_LABEL: &str = "serialization";
 pub const DESERIALIZATION_LABEL: &str = "deserialization";
 
-pub static APTOS_CONNECTIONS: Lazy<IntGaugeVec> = Lazy::new(|| {
-    register_int_gauge_vec!(
-        "aptos_connections",
-        "Number of current connections and their direction",
-        &["role_type", "network_id", "peer_id", "direction"]
+// previous APTOS_CONNECTIONS counter replaced by PeersAndMetadataGauge which translates PeersAndMetadata global singleton directly to Prometheus Metrics data
```
Comment (JoshLind) [inline network/framework/src/counters.rs:44]: Does it make sense for us to create a new dashboard containing the metrics for the new code? It's hard to track via the diff on what's been removed/broken/changed.

File: network/framework/src/peer.rs [hunk 26]
```diff
@@ -0,0 +1,816 @@
...
+            reason = close_reason,
+            op = "writer_thread closing",
+            peer = self.peer_network_id,
+            "peerclose"
+        );
+        closed.close().await;
+    }
+}
+
+pub static NETWORK_PEER_MESSAGE_FRAMES_WRITTEN: Lazy<IntCounterVec> = Lazy::new(||
```
Comment (JoshLind) [inline network/framework/src/peer.rs:340]: Do we want to keep these counters in this file, or move them to the dedicated file? Probably makes sense to be consistent and either use a single file, or move all the counters to appropriate places.

File: consensus/src/block_storage/tracing.rs (hunk 21 above)
Comment (brianolson) [reply consensus/src/block_storage/tracing.rs:23]: unused. commenting it out was just a double check that everything compiles without it. find-all shows no other 'rand_agg_decision'

File: config/src/config/network_config.rs (hunk 20 above)
Comment (brianolson) [reply config/src/config/network_config.rs:348]: https://github.com/aptos-labs/aptos-core/pull/12638

File: consensus/src/block_storage/tracing.rs (hunk 21 above)
Comment (brianolson) [reply consensus/src/block_storage/tracing.rs]: later work on `main` made this a merge conflict because they fixed it there.

File: network/framework/src/application/storage.rs (hunk 22 above)
Comment (brianolson) [reply network/framework/src/application/storage.rs]: Yup. Did that, but then replaced it all with the ArcSwap stuff from `main`

File: network/framework/src/counters.rs (hunk 25 above)
Comment (brianolson) [reply network/framework/src/counters.rs:44]: There's a new section in the `network` dashboard for net2 stuff. We can add misc new graphs there.

File: aptos-node/src/lib.rs (hunk 19 above)
Comment (brianolson) [reply aptos-node/src/lib.rs:635]: Some of this refactor relies on the internal changes. I don't think it makes sense in the current code.
